bin/generate_tests.py

```diff
@@ -0,0 +1,103 @@
+#!/usr/bin/env python3.7
+import argparse
+import json
+import logging
+import os
+import re
+import sys
+import tempfile
+from glob import glob
+from subprocess import check_call, CalledProcessError, DEVNULL
+
+from jinja2 import Environment, FileSystemLoader, TemplateNotFound
+
+VERSION = '0.1.0'
+
+DEFAULT_SPEC_LOCATION = os.path.join('..', 'problem-specifications')
+RGX_WORDS = re.compile(r'[-_\s]|(?=[A-Z])')
+
+logging.basicConfig()
+logger = logging.getLogger('generator')
+logger.setLevel(logging.WARN)
+
+
+def snake_case(string):
+    return '_'.join(RGX_WORDS.split(string)).lower()
+
+
+def title_case(string):
+    return''.join(word.title() for word in RGX_WORDS.split(string))
+
+
+def load_canonical(exercise, spec_path):
+    full_path = os.path.join(
+        spec_path, 'exercises', exercise, 'canonical-data.json'
+    )
+    with open(full_path) as f:
+        return json.load(f)
+
+
+def format_file(path):
+    check_call(['black', '-q', path])
+
+
+def generate_exercise(env, spec_path, exercise, check=False):
+    slug = os.path.basename(exercise)
+    try:
+        spec = load_canonical(slug, spec_path)
+        template_path = os.path.join(slug, '.meta', 'template.j2')
+        try:
+            template = env.get_template(template_path)
+            tests_path = os.path.join(
+                exercise, f'{snake_case(slug)}_test.py'
+            )
+            rendered = template.render(**spec)
+            if check:
+                tmp_f, tmp_path = tempfile.mkstemp()
```

Comment:
Seems to me we could do the diff without resort to subprocess, if we're just looking to see if the files would match.

Reply:
I was inclined to think the `diff` tool would be faster at comparing whole files... although they are fairly small files, even for the larger exercises.

Reply:
The overhead of the subprocess will trump that until the files get _big_ ... if we're looking for _identical_ content then just opening any existing file and comparing content is sufficient.